File: diquis-local-backend/app/schemas/transaction.py
from pydantic import BaseModel, Field
from typing import Optional, List
from uuid import UUID
from datetime import datetime

class TransactionCreate(BaseModel):
    type: str = Field(..., pattern="^(income|expense|transfer)$", description="Debe ser 'income', 'expense' o 'transfer'")
    amount: float = Field(..., gt=0, description="El monto debe ser mayor a 0")
    account_id: Optional[UUID] = None
    transfer_to_account_id: Optional[UUID] = None
    category_id: Optional[UUID] = None
    description: Optional[str] = None
    occurred_at: Optional[datetime] = None
    currency: Optional[str] = "CRC"
    payment_method: Optional[str] = Field("cash", pattern="^(cash|debit|credit|transfer|sinpe|digital)$")
    is_essential: Optional[bool] = False
    notes: Optional[str] = None
    exchange_rate: Optional[float] = Field(
        None, description="Tasa de cambio aplicada (override del usuario)"
    )
    exchange_source: Optional[str] = Field(
        "none", description="De dónde salió la tasa: 'user', 'api', 'bccr'"
    )

class TransactionResponse(BaseModel):
    id: UUID
    type: str
    amount: float
    currency: str
    amount_in_account_currency: float
    exchange_rate: float
    exchange_source: str
    description: Optional[str]
    category_id: Optional[UUID]
    account_id: UUID
    transfer_to_account_id: Optional[UUID]
    occurred_at: datetime
    is_essential: bool
    payment_method: str
    notes: Optional[str]
    is_deleted: bool

    class Config:
        from_attributes = True

# --- MODELOS PARA EL BALANCE ---
class BalanceMetrics(BaseModel):
    monthly_income: float
    monthly_expense: float

class BalanceResponse(BaseModel):
    wallet_name: str
    currency: str
    current_balance: float
    metrics: BalanceMetrics

# --- MODELOS PARA EL DETALLE (ÍTEMS) ---
class TransactionItemCreate(BaseModel):
    product_name: str = Field(..., min_length=1, description="Nombre del producto")
    subcategory_id: Optional[UUID] = None
    quantity: float = Field(1.0, gt=0, description="Cantidad comprada")
    line_total: float = Field(..., ge=0, description="Monto total pagado por esta línea")
    is_essential: bool = False

class TransactionItemResponse(BaseModel):
    id: UUID
    transaction_id: UUID
    product_name: str
    subcategory_id: Optional[UUID]
    quantity: float
    unit_price: float
    line_total: float
    is_essential: bool

    class Config:
        from_attributes = True

class TransactionDetailResponse(TransactionResponse):
    items: List[TransactionItemResponse] = []

    class Config:
        from_attributes = True

class TransactionUpdate(BaseModel):
    category_id: Optional[UUID] = None
    account_id: Optional[UUID] = None
    transfer_to_account_id: Optional[UUID] = None
    amount: Optional[float] = Field(None, gt=0)
    description: Optional[str] = None
    occurred_at: Optional[datetime] = None
    type: Optional[str] = Field(None, pattern="^(income|expense|transfer)$")
    payment_method: Optional[str] = Field(None, pattern="^(cash|debit|credit|transfer|sinpe|digital)$")
    is_essential: Optional[bool] = None
    notes: Optional[str] = None
    exchange_rate: Optional[float] = Field(None, description="Permite corregir la tasa aplicada")
    exchange_source: Optional[str] = Field(None, description="Actualizar el origen si es necesario")

File: diquis-local-frontend/src/types/transaction.ts
export type PaymentMethod = 'cash' | 'debit' | 'credit' | 'transfer' | 'sinpe' | 'digital';
export type TransactionType = 'income' | 'expense' | 'transfer';

export interface BalanceMetrics {
    monthly_income: number;
    monthly_expense: number;
}

export interface BalanceData {
    wallet_name: string;
    currency: string;
    current_balance: number;
    metrics: BalanceMetrics;
}

export interface TransactionItem {
    id: string;
    transaction_id: string;
    product_name: string;
    subcategory_id: string | null;
    quantity: number;
    unit_price: number;
    line_total: number;
    is_essential: boolean;
}

export interface Transaction {
    id: string;
    type: TransactionType;
    amount: number;
    currency: string;
    amount_in_account_currency: number;
    exchange_rate: number;
    exchange_source: string;
    description: string | null;
    category_id: string | null;
    account_id: string;
    transfer_to_account_id: string | null;
    occurred_at: string;
    is_essential: boolean;
    payment_method: PaymentMethod;
    notes: string | null;
    is_deleted: boolean;
}

export interface TransactionDetail extends Transaction {
    items: TransactionItem[];
}

export interface TransactionCreate {
    type: TransactionType;
    amount: number;
    account_id?: string;
    transfer_to_account_id?: string;
    category_id?: string;
    description?: string;
    occurred_at?: string;
    payment_method?: PaymentMethod;
    notes?: string;
    is_essential?: boolean;
    currency?: string;
    exchange_rate?: number;
    exchange_source?: string;
}

export interface TransactionUpdate {
    type?: TransactionType;
    amount?: number;
    category_id?: string;
    account_id?: string;
    transfer_to_account_id?: string;
    description?: string;
    occurred_at?: string;
    payment_method?: PaymentMethod;
    notes?: string;
    exchange_rate?: number;
    exchange_source?: string;
}

export interface TransactionItemCreate {
    product_name: string;
    subcategory_id?: string;
    quantity: number;
    line_total: number;
    is_essential: boolean;
}